core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java

```diff
@@ -0,0 +1,210 @@
+    // by the dependencies between nodes
+    LinkedHashMap<ID, Set<ID>> directPredecessors = new LinkedHashMap<>(nodes.size());
+    Map<ID, Set<ID>> directSuccessors = new HashMap<>();
+
+    List<ID> results = new ArrayList<>(nodes.size());
+
+    for (N node : nodes) {
+      ID nodeID = identityExtractor.apply(node);
+      // we need the set to be modifiable, so let's make our own
+      Set<ID> preds = new HashSet<>(directPredecessorsExtractor.apply(node));
+      directPredecessors.put(nodeID, preds);
+      for (ID pred : preds) {
+        directSuccessors.computeIfAbsent(pred, __ -> new HashSet<>()).add(nodeID);
+      }
+      idMap.put(nodeID, node);
+    }
+
+    sort(directPredecessors, directSuccessors, results);
+
+    return results.stream().map(idMap::get).collect(Collectors.toList());
+  }
+
+  private void sort(
+      LinkedHashMap<ID, Set<ID>> predecessors, Map<ID, Set<ID>> successors, List<ID> results) {
+
+    while (!predecessors.isEmpty()) {
+      ID curr = removeFirstIndependent(predecessors, successors);
+      if (curr != null) {
+        // yay, simple. Just add the found independent node to the results.
+        results.add(curr);
+      } else {
+        // ok, there is a cycle in the graph. Let's remove all the nodes in the first cycle we find
+        // from our predecessors map, add them to the result in their original order and try to
+        // continue normally
+
+        // find the first cycle in the predecessors (in the original list order)
+        Iterator<ID> nexts = predecessors.keySet().iterator();
+        List<ID> cycle;
+        do {
+          curr = nexts.next();
+          cycle = findCycle(curr, predecessors);
+        } while (cycle.isEmpty() && nexts.hasNext());
+
+        // If we ever find a graph that doesn't have any independent node, yet we fail to find a
+        // cycle in it, the universe must be broken.
+        if (cycle.isEmpty()) {
+          throw new IllegalStateException(
+              String.format(
+                  "Failed to find a cycle in a graph that doesn't seem to  have any independent"
+                      + " node. This should never happen. Please file a bug. Current state of the"
+                      + " sorting is: predecessors=%s, successors=%s, results=%s",
+                  predecessors.toString(), successors.toString(), results.toString()));
+        }
+
+        // remove the cycle from the predecessors and update the predecessor map so that dependent
+        // nodes that are not part of the cycle can be sorted properly
+        removePredecessorMapping(predecessors, successors, curr);
+        results.add(curr);
+
+        // make sure the cycle doesn't contain our current node so that it doesn't appear in the
+        // results twice
+        cycle.remove(curr);
+
+        for (ID n : cycle) {
+          removePredecessorMapping(predecessors, successors, n);
```

Comment:
You can avoid duplicating this code 
```java
removePredecessorMapping(predecessors, successors, curr);
        results.add(curr);
```
Like the following
```java
...
        TreeSet<ID> cycle;
...
        // If we ever find a graph that doesn't have any independent node, yet we fail to find a
        // cycle in it, the universe must be broken.
        if (cycle.isEmpty()) {
          throw new IllegalStateException(
              String.format(
                  "Failed to find a cycle in a graph that doesn't seem to  have any independent"
                      + " node. This should never happen. Please file a bug. Current state of the"
                      + " sorting is: predecessors=%s, successors=%s, results=%s",
                  predecessors.toString(), successors.toString(), results.toString()));
        }

        // remove the cycle from the predecessors and update the predecessor map so that dependent
        // nodes that are not part of the cycle can be sorted properly
        // cycle elements are inserted in result in order they were found but not original order
        for(ID id : cycle) {
          removePredecessorMapping(predecessors, successors, id);
          results.add(id);
        }
      }
    }
  }

  private void removePredecessorMapping(
      Map<ID, Set<ID>> predecessors, Map<ID, Set<ID>> successors, ID node) {
    for (ID succ : successors.getOrDefault(node, emptySet())) {
      predecessors.getOrDefault(succ, emptySet()).remove(node);
    }
    predecessors.remove(node);
  }

  private TreeSet<ID> findCycle(ID node, Map<ID, Set<ID>> predecessors) {
    // bail out quickly if there are no preds - should be fairly common occurrence hopefully
    Set<ID> preds = predecessors.get(node);
    if (preds == null || preds.isEmpty()) {
      return new TreeSet<>();
    }

    TreeSet<ID> ret = new TreeSet<>();
    // start node is first item of result
    ret.add(node);
    List<ID> todo = new ArrayList<>(preds);

    Set<ID> visited = new HashSet<>();

    while (!todo.isEmpty()) {
      ID n = todo.remove(0);

      if (visited.contains(n)) {
        continue;
      }

      visited.add(n);

      todo.addAll(predecessors.getOrDefault(n, emptySet()));
      ret.add(n);

      if (n.equals(node)) {
        // we found the cycle to our original node
        return ret;
      }
    }

    return new TreeSet<>();
  }
...
```
WDYT?